File: leetcode/lc378.cpp
#include <vector>
#include <algorithm>
#include <cassert>

using namespace std;

class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        assert(k >= 0);
        const size_t nrows = matrix.size(),
                     ncols = matrix[0].size();
        size_t kk = k;

        int low = matrix[0][0], high = matrix[nrows-1][ncols-1];
        int mid;

        while (low <= high) {
            mid = int((long(low) + long(high)) / 2);
            size_t count = 0;
            for (size_t i = 0; i < nrows; ++i) {
                auto row = matrix[i];
                count += upper_bound(row.begin(), row.end(), mid) - row.begin();
            }
            if (count >= kk)
                high = mid - 1;
            else if (count < kk)
                low = mid + 1;
        }

        return low;
    }
};

File: leetcode/lc378.py
from heapq import heappush, heappop
# O(k log k) solution.
class Solution(object):
    def kthSmallest(self, matrix, k):
        """
        :type matrix: List[List[int]]
        :type k: int
        :rtype: int
        """
        assert k >= 1
        if k == 1:
            return matrix[0][0]
        nrows = len(matrix)
        ncols = len(matrix[0])
        visited = {(0, 0)}
        heap = [(matrix[0][0], 0, 0)]
        discarded = 0
        
        while discarded != k - 1:
            _, row, col = heappop(heap)
            discarded += 1
            if row + 1 < nrows and (row + 1, col) not in visited:
                heappush(heap, (matrix[row+1][col], row + 1, col))
                visited.add((row + 1, col))
            if col + 1 < ncols and (row, col + 1) not in visited:
                heappush(heap, (matrix[row][col+1], row, col + 1))
                visited.add((row, col + 1))

        return heap[0][0]
